src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1976,6 +1977,58 @@ multi_relation_restriction_hook(PlannerInfo *root, RelOptInfo *relOptInfo,
 }
 
 
+/*
+ * Normally, we should not need this. However, the combination of
+ * Postgres commit 3c569049b7b502bb4952483d19ce622ff0af5fd6 and
+ * Citus function AdjustPartitioningForDistributedPlanning()
+ * forces us to do this. The commit expects partitioned indexes
+ * to belong to relations with "inh" flag set properly. Whereas, the
+ * function overrides "inh" flag. To avoid a crash,
+ * we go over the list of indexinfos and remove all partitioned indexes.
+ * Partitioned indexes were ignored pre PG16 anyway, we are essentially
+ * not breaking any logic.
+ *
+ * AdjustPartitioningForDistributedPlanning() is a hack that we use
+ * to prevent Postgres' standard_planner() to expand all the partitions
+ * for the distributed planning when a distributed partitioned table
+ * is queried. It is required for both correctness and performance
+ * reasons. Although we can eliminate the use of the function for
+ * the correctness (e.g., make sure that rest of the planner can handle
+ * partitions), it's performance implication is hard to avoid. Certain
+ * planning logic of Citus (such as router or query pushdown) relies
+ * heavily on the relationRestrictionList. If
+ * AdjustPartitioningForDistributedPlanning() is removed, all the
+ * partitions show up in the, causing high planning times for
+ * such queries.
+ */
+void
+multi_partitioned_index_hook(PlannerInfo *root, Oid relationObjectId, bool inhparent,
+							 RelOptInfo *rel)
+{
+	Index varno = rel->relid;
+	RangeTblEntry *rangeTableEntry = planner_rt_fetch(varno, root);
+
+	if (rangeTableEntry->rtekind == RTE_RELATION &&
```

Comment:
do we need something like?
```C
	if (!CitusHasBeenLoaded() || !CheckCitusVersion(DEBUG5))
	{
		return;
	}
```

Also, why do we do this for all tables, not Citus tables? Probably should use `IsCitusTable` at some point.

 `IsCitusTable` already does the `CitusHasBeenLoadedCheck`?

Reply:
Actually, we do this for all tables, not just Citus tables. `AdjustPartitioningForDistributedPlanning` function doesn't filter on Citus tables.
Adding CitusHasBeenLoaded check makes sense, thanks.

Reply:
Also, if it's not fastPathRouterQuery. needsDistributedPlanning also requires traversing the range table list. This is the code, and based on the following I decided that it's easier to just not check whether AdjustDistributedPlanning has been done through these flags:

```C
List *rangeTableList = ExtractRangeTableEntryList(parse);

	if (cursorOptions & CURSOR_OPT_FORCE_DISTRIBUTED)
	{
		/* this cursor flag could only be set when Citus has been loaded */
		Assert(CitusHasBeenLoaded());

		needsDistributedPlanning = true;
	}
	else if (CitusHasBeenLoaded())
	{
		needsDistributedPlanning =
			ListContainsDistributedTableRTE(rangeTableList,
											&maybeHasForeignDistributedTable);
		if (needsDistributedPlanning)
		{
			fastPathRouterQuery = FastPathRouterQuery(parse, &distributionKeyValue);
		}
	}

	if (needsDistributedPlanning)
	{
		if (!fastPathRouterQuery)
		{
			/*
			 * When there are partitioned tables (not applicable to fast path),
			 * pretend that they are regular tables to avoid unnecessary work
			 * in standard_planner.
			 */
			bool setPartitionedTablesInherited = false;
			AdjustPartitioningForDistributedPlanning(rangeTableList,
													 setPartitionedTablesInherited);
		}
	}
```

Reply:
Since `AdjustPartitioningForDistributedPlanning` also adjusts regular Postgres tables, I need to call this hook for regular Postgres tables as well. The only condition I can put is `CitusHasBeenLoaded`. I cannot put `IsCitusTable` condition - it's not sufficient!

What I could put is `needsDistributedPlanning` combined with `!fastPathRouterQuery` condition, which will give me the exact case in which `AdjustPartitioningForDistributedPlanning` function is called, but as I mentioned above that condition is not the simplest to get.

I think it's good enough to check if Citus has been loaded, and then check if the table is partitioned but its inh flag is false - this means `AdjustPartitioningForDistributedPlanning` function is called, because there is no scenario in plain Postgres where this is true.

Reply:
> I think it's good enough to check if Citus has been loaded, and then check if the table is partitioned but its inh flag is false - this means AdjustPartitioningForDistributedPlanning function is called, because there is no scenario in plain Postgres where this is true.

I could add also add a check for `ListContainsDistributedTableRTE`, what do you think?

Reply:
> I think it's good enough to check if Citus has been loaded, and then check if the table is partitioned but its inh flag is false - this means AdjustPartitioningForDistributedPlanning function is called, because there is no scenario in plain Postgres where this is true.

So, maybe we should add a new function: `PartitionedRTEWentThroughAdjustPartitioning()` which simply checks this. Then, it is much better documented.

Then, I can accept this solution.

Reply:
Hmm, good idea, let me do that.